### ISSUE TITLE
[DI] Decorate the service directly instead of override

### ISSUE THREAD

**Issue by rapisarda** [symfony/symfony#30599]
**Description**
Add possibility to decorate the service directly instead of override.

**Example**

override:

```yaml
services:
    ham_burger:
        class: Burger

    cheese_burger:
        class: Burger

    ham.ham_decorator:
        class: HamDecorator
        decorate: ham_burger

    cheese.cheese_decorator:
        class: CheeseDecorator
        decoration_priority: 2
        decorate: cheese_burger

    ham.cheese_decorator:
        class: CheeseDecorator
        decoration_priority: 1
        decorate: cheese_burger
```

decorate :

```yaml
services:
    ham_burger:
        class: Burger
        decorators:
            - HamDecorator

    cheese_burger:
        class: Burger
        decorators:
            - CheeseDecorator
            - HamDecorator
```

**Comment by ro0NL** [comment]
At first i liked it :) but we introduce a new (less flexible) concept i believe.

In the first case you are free to modify each decorator as needed. In your proposal we lose this possibility. If we make e.g. `decorators: [id, ..., idN]` we're back to the current situation more or less by referencing existing services (you could configure them with `decorate`).

TLDR; im not sure how much it increases the complexity of service configuration, given it opens multiple paths.

**Comment by stof** [comment]
your proposal forbids configuring the decorator services. So it is inferior to the current solution.

**Comment by rapisarda** [comment]
I don't understand what forbids configuring the decorator services either what is less flexible

```yaml
services:
    ham_burger:
        class: Burger
        autowire: true
        decorators:
            - HamDecorator

    cheese_burger:
        class: Burger
        decorators:
            - CheeseDecorator
            - HamDecorator
        arguments:
            - 'foo'

  CheeseDecorator:
    arguments: 
      - '@CheeseDecorator.inner'
      - 'cheddar'

  ketchupDecorator:
    decorate: '@ham_burger'
```

**Comment by ro0NL** [comment]
it shouldnt be required to use class named service IDs

My worry is `CheeseDecorator` becomes an implicit decorator (you wouldnt know, only if you see the corresponding cheese_burger service).

**Comment by rapisarda** [comment]
It's ok,  CheeseDecorator definition is ugly

Class named service IDs is mandatory if you want a lot of configuration for the same class and this is probably in case you implement decorator pattern

Thanks for your feedback

**Comment by ro0NL** [comment]
> Class named service IDs is mandatory if you want a lot of configuration for the same class and this is probably in case you implement decorator pattern

not sure i understand that, the same class can be used as a decorator N times.

**Comment by stof** [comment]
and it also requires the service being decorated to know about its decorators, so it is also less flexible as an extension point.

**Comment by stof** [comment]
Well, in your proposal, things indeed break if 2 services declare `HamDecorator` as being their decorator as they would refer to the same service, and so `HamDecorator.inner` would be one of them.

**Comment by rapisarda** [comment]
What if `decorators: [CheeseDecorator]` make a decorator service named like "cheese_burger.CheeseDecorator.decorator" which extend CheeseDecorator service which can be previously declared as abstract ?
And `decorators: {CheeseDecorator: fooDecorator}` make a decorator service named "fooDecorator" ?

**Comment by stof** [comment]
Well, the fact that this would have to create other services would make it a nightmare to configure them if there is any need for an explicit configuration.
And it still does not change the fact that having a service knowing about all its decorator at config time is a much less powerful extension point.

**Comment by psren** [comment]
Is your proposal just syntactical sugar for smth like that (which is already possible)?

```yaml
services:
    burger:
        class: Burger
        arguments:
            - !service
                class: Ham
                arguments:
                    - !service
                        class: Cheese
```

**Comment by rapisarda** [comment]
Exactly. In my case i have something like 30 decorator service with something like 15 decorated services

**Comment by nicolas-grekas** [comment]
This could work and would be pretty neat to me:
```yaml
services:
    burger:
        stack:
            - SomeClass: {arguments: [@stack.next]}
            - AnotherClass:
                calls: [setDecorated: ['@stack.next']]
			- TheLastClass # string <=> TheLastClass: []
```

This would be turned into actual decorators by loaders.
